viewer/src/models/cad/CadModelImpl.ts

```diff
@@ -0,0 +1,107 @@
+/*!
+ * Copyright 2020 Cognite AS
+ */
+
+import { createParser, createQuadsParser } from './parseSectorData';
+import { ParseSectorDelegate } from './delegates';
+import { CadModel } from './CadModel';
+import { SectorScene, SectorModelTransformation, SectorQuads, Sector } from './types';
+import { ModelDataRetriever } from '../../datasources/ModelDataRetriever';
+import { CadMetadataParser } from './CadMetadataParser';
+
+export class CadModelImpl implements CadModel {
+  private _scene?: SectorScene;
+  private readonly _modelTransformation: SectorModelTransformation;
+
+  private readonly dataRetriever: ModelDataRetriever;
+  private readonly detailedParser: ParseSectorDelegate<Sector>;
+  private readonly simpleParserPromise: Promise<ParseSectorDelegate<SectorQuads>>;
+  private readonly scenePromise: Promise<SectorScene>;
+
+  /**
+   * Do not use directly, see loadCadModelByUrl().
+   * @see loadCadModelByUrl
+   */
+  constructor(dataRetriever: ModelDataRetriever, modelTransformation: SectorModelTransformation) {
+    this.dataRetriever = dataRetriever;
+    this.detailedParser = createParser(id => this.fetchCtm(id));
+    this.simpleParserPromise = createQuadsParser();
+    this._modelTransformation = modelTransformation;
+
+    const metadataJson = this.dataRetriever.fetchJson('scene.json');
+    this.scenePromise = parseSceneMetadata(metadataJson);
+    this._scene = undefined;
+  }
```

Comment:
Make constructor private and create a static async create function. Then you can remove the initialize function and all the ensureInitialized calls:

```
export class CadModelImpl implements CadModel {
  static async create(
    dataRetriever: ModelDataRetriever,
    modelTransformation: SectorModelTransformation
  ): Promise<CadModelImpl> {
    const metadataJson = dataRetriever.fetchJson('scene.json');
    const scene = await parseSceneMetadata(metadataJson);
    const model = new CadModelImpl(dataRetriever, modelTransformation, scene);
    return model;
  }

  private readonly _scene: SectorScene;
  private readonly _modelTransformation: SectorModelTransformation;

  private readonly dataRetriever: ModelDataRetriever;
  private readonly detailedParser: ParseSectorDelegate<Sector>;
  private readonly simpleParserPromise: Promise<ParseSectorDelegate<SectorQuads>>;

  private constructor(
    dataRetriever: ModelDataRetriever,
    modelTransformation: SectorModelTransformation,
    scene: SectorScene
  ) {
    this.dataRetriever = dataRetriever;
    this.detailedParser = createParser(id => this.fetchCtm(id));
    this.simpleParserPromise = createQuadsParser();
    this._modelTransformation = modelTransformation;
    this._scene = scene;
  }
  // ...
}
```

In `loadCadModelByUrl` you can now call

```
  const model = await CadModelImpl.create(retriever, transform);
```
This makes sure `_scene` is always initialized and the `CadModelImpl` is never in an incomplete state.

Reply:
🥇 Fixed